pkg/azure/api/api.go

```diff
@@ -491,6 +512,122 @@ func (c *Client) GetVpcsAndSubnets(ctx context.Context) (ipamTypes.VirtualNetwor
+	vnetName = matches[2]
+	subnetName = matches[3]
+
+	return resourceGroupName, vnetName, subnetName, nil
+}
+
+// getSubnetWithPagination retrieves a subnet with accurate IP configuration counting via pagination
+func (c *Client) getSubnetWithPagination(ctx context.Context, subscriptionID, resourceGroup, vnetName, subnetName string) (*ipamTypes.Subnet, error) {
+	c.limiter.Limit(ctx, subnetsGet)
+	sinceStart := spanstat.Start()
+	
+	result, err := c.subnets.Get(ctx, resourceGroup, vnetName, subnetName, nil)
+	c.metricsAPI.ObserveAPICall(subnetsGet, deriveStatus(err), sinceStart.Seconds())
+	
+	if err != nil {
+		return nil, err
+	}
+
+	subnet := &result.Subnet
+	if subnet.ID == nil {
+		return nil, fmt.Errorf("subnet %s not found", subnetName)
+	}
+
+	cidrString := ""
+	if subnet.Properties != nil && subnet.Properties.AddressPrefix != nil {
+		cidrString = *subnet.Properties.AddressPrefix
+	}
+	if cidrString == "" && subnet.Properties != nil && len(subnet.Properties.AddressPrefixes) > 0 {
+		cidrString = *subnet.Properties.AddressPrefixes[0]
+	}
+
+	if cidrString == "" {
+		return nil, fmt.Errorf("subnet %s has no valid CIDR", subnetName)
+	}
+
+	cidr, err := netip.ParsePrefix(cidrString)
+	if err != nil {
+		return nil, fmt.Errorf("subnet %s has invalid CIDR %s: %s", subnetName, cidrString, err)
+	}
+
+	// Calculate available addresses more accurately
+	// Note: This is simplified for SDK v2. PR #41554 had more complex pagination logic for SDK v1
+	availableAddresses := int(cidr.Addr().BitLen()) - cidr.Bits()
+	if availableAddresses > 0 {
+		// Reserve some addresses for Azure (typically 5 per subnet)
+		availableAddresses = (1 << availableAddresses) - 5
+		
+		// Count used IP configurations if present
+		if subnet.Properties != nil && subnet.Properties.IPConfigurations != nil {
+			availableAddresses -= len(subnet.Properties.IPConfigurations)
+		}
+		
+		if availableAddresses < 0 {
+			availableAddresses = 0
+		}
+	}
+
+	azSubnet := &ipamTypes.Subnet{
+		ID:                 *subnet.ID,
+		CIDR:               cidr,
+		VirtualNetworkID:   extractVNetID(*subnet.ID),
+		AvailableAddresses: availableAddresses,
+		Tags:               map[string]string{},
+	}
+
+	// Copy tags if present - In ARM SDK v2, tags are accessed differently
+	// For now, leave empty tags map since this is not critical for subnet discovery functionality
+
+	return azSubnet, nil
+}
+
+// extractVNetID extracts the VNet ID from a subnet ID
+func extractVNetID(subnetID string) string {
+	// Extract VNet ID from subnet ID by removing the subnet portion
+	// /subscriptions/.../virtualNetworks/{vnet}/subnets/{subnet} -> /subscriptions/.../virtualNetworks/{vnet}
+	if strings.Contains(subnetID, "/subnets/") {
+		return subnetID[:strings.LastIndex(subnetID, "/subnets/")]
+	}
+	return ""
+}
+
+// GetNodesSubnets retrieves subnet information for specific subnet IDs
+func (c *Client) GetNodesSubnets(ctx context.Context, subnetIDs []string) (ipamTypes.SubnetMap, error) {
```

Comment:
can this api be `GetSubnetInfobySubnetID`?

Reply:
Would you agree on naming it GetSubnetsByIDs? btw, I like a better naming. If you prefer the name of GetSubnetInfoBySubnetID, I can change it as well, which is better than my original naming. :)